pkg/agent/cache/manager.go

```diff
@@ -0,0 +1,382 @@
+package cache
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+
+	"crypto/tls"
+	"crypto/x509"
+	"github.com/sirupsen/logrus"
+	spiffe_tls "github.com/spiffe/go-spiffe/tls"
+	"github.com/spiffe/go-spiffe/uri"
+	"github.com/spiffe/spire/pkg/common/util"
+	"github.com/spiffe/spire/proto/api/node"
+	proto "github.com/spiffe/spire/proto/common"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"io"
+	"sync"
+	"time"
+)
+
+type Manager interface {
+	Init()
+	Cache() Cache
+	Shutdown(err error)
+	Done() <-chan struct{}
+	Err() error
+}
+
+type EntryRequest struct {
+	CSR   []byte
+	entry CacheEntry
+}
+
+type MgrConfig struct {
+	ServerCerts    []*x509.Certificate
+	ServerSPIFFEID string
+	ServerAddr     string
+	BaseSVID       []byte
+	BaseSVIDKey    *ecdsa.PrivateKey
+	BaseRegEntries []*proto.RegistrationEntry
+	Logger         logrus.FieldLogger
+}
+
+type manager struct {
+	managedCache     Cache
+	serverCerts      []*x509.Certificate
+	serverAddr       string
+	serverSPIFFEID   string
+	reason           error
+	entryRequestCh   chan map[string][]EntryRequest
+	regEntriesCh     chan []*proto.RegistrationEntry
+	cacheEntryCh     chan CacheEntry
+	spiffeIdEntryMap map[string]CacheEntry
+	baseSVID         []byte
+	baseSVIDKey      *ecdsa.PrivateKey
+	baseSPIFFEID     string
+	baseRegEntries   []*proto.RegistrationEntry
+	log              logrus.FieldLogger
+	ctx              context.Context
+	cancel           context.CancelFunc
+	doneCh           chan struct{}
+}
+
+func NewManager(ctx context.Context, c *MgrConfig) (Manager, error) {
+
+	cert, err := x509.ParseCertificate(c.BaseSVID)
+	if err != nil {
+		return nil, err
+	}
+	basespiffeID, err := uri.GetURINamesFromCertificate(cert)
+	if err != nil {
+		return nil, err
+	}
+	ctx, cancel := context.WithCancel(ctx)
+
+	return &manager{
+		managedCache:     NewCache(c.Logger),
+		serverCerts:      c.ServerCerts,
+		serverSPIFFEID:   c.ServerSPIFFEID,
+		serverAddr:       c.ServerAddr,
+		baseSVID:         c.BaseSVID,
+		baseSVIDKey:      c.BaseSVIDKey,
+		baseSPIFFEID:     basespiffeID[0],
+		baseRegEntries:   c.BaseRegEntries,
+		log:              c.Logger.WithField("subsystem_name", "cacheManager"),
+		spiffeIdEntryMap: make(map[string]CacheEntry),
+		entryRequestCh:   make(chan map[string][]EntryRequest),
+		regEntriesCh:     make(chan []*proto.RegistrationEntry),
+		cacheEntryCh:     make(chan CacheEntry),
+		ctx:              ctx,
+		cancel:           cancel,
+		doneCh:           make(chan struct{}),
+	}, nil
+}
+
+func (m *manager) Shutdown(err error) {
+	m.log.Debug("Shutting Down Cache Manager ", err)
+	m.reason = err
```

Comment:
There's a race condition here because `m.Shutdown()` is called from multiple threads.

Also, every `Shutdown(err)` will overwrite the previous `reason`.  The initial error is generally more useful imho.

A quick fix/hack would be to wrap it in a [`sync.Once`](https://golang.org/pkg/sync/#Once):

```go
m.once.Do(func(){
   m.reason = err
})
```

Reply:
created a errch channel, the errch receiver would drain all errors and only call shutdown with the first error.